networking_cisco/plugins/cisco/db/l3/l3_router_appliance_gbp_db.py

```diff
@@ -0,0 +1,130 @@
+                    fip['subnet_id'] = nat_pool['subnet_id']
+                    try:
+                        result = super(l3_r_db.L3RouterApplianceDBMixin,
+                                    self).create_floatingip(context,
+                                                            floatingip)
+                        router_ids = ([result['router_id']]
+                                      if result['router_id'] else [])
+                    except n_exc.IpAddressGenerationFailure as ex:
+                        LOG.warning(_LW("Floating allocation failed: %s"),
+                                    ex.message)
+                    if result:
+                        break
+        if not result:
+            info = super(l3_r_db.L3RouterApplianceDBMixin,
+                         self).create_floatingip(context,
+                                                 floatingip, initial_status)
+            router_ids = [info['router_id']] if info['router_id'] else []
+        context.result = result
+        if driver:
+            driver.create_floatingip_postcommit(context, fip_ctxt)
+        self._notify_affected_routers(context, router_ids, 'create_floatingip')
+        return result
+
+    def update_floatingip(self, context, floatingip_id, floatingip):
+        """Specialization of parent class's update_floatingip.
+
+        This specialization of the parent class's update_floatingip
+        is needed for GBP workflow in order to pass the result to the
+        postcommit call.
+        """
+
+        orig_fl_ip = super(l3_r_db.L3RouterApplianceDBMixin, self).get_floatingip(
+            context, floatingip_id)
+        before_router_id = orig_fl_ip['router_id']
+        if before_router_id:
+            router_type_id = self.get_router_type_id(context, before_router_id)
+            driver = self._get_router_type_driver(context,
+                                                  router_type_id)
+            if driver:
+                fip_ctxt = driver_context.FloatingipContext(
+                        floatingip.get('floatingip'), orig_fl_ip)
+                driver.update_floatingip_precommit(context, fip_ctxt)
+        else:
+            fip_ctxt = None
+
+        info = super(l3_r_db.L3RouterApplianceDBMixin, self).update_floatingip(
+            context, floatingip_id, floatingip)
+        router_ids = []
+        if before_router_id:
+            router_ids.append(before_router_id)
+        r_id = info['router_id']
+        if r_id and r_id != before_router_id:
+            router_ids.append(r_id)
+        if r_id:
+            router_type_id = self.get_router_type_id(context, r_id)
+            driver = self._get_router_type_driver(context,
+                                                  router_type_id)
+            if not fip_ctxt:
+                fip_ctxt = driver_context.FloatingipContext(
+                    floatingip.get('floatingip'), orig_fl_ip)
+            if driver:
+                context.result = info
```

Comment:
you can extract all the code in the update_floatingip() of the parent's class into a helper function but with a new parameter flag (default to False) to control the adding of this extra line before passing it to postcommit(). Then in the update_floatingip() of the parent's class it will just call this helper. In the update_floatingip() of your class, you will just call the helper too but with the flag setting it to True.